### PR TITLE
ui: discover latency in very noisy environments

### PR DESCRIPTION
**Pull request by hunchback**
We are attempting to calculate the latency between A to B by finding a flow
which has been seen in both points. Due to current limitions in Gremlin we
are not able to do so in a single query forcing us to:
- 1st find a flow which intersects with A (using a small queue)
- 2nd check if the discovered flow intersects with B
If the system is wieghted down with many unique flows running from A to C we
may fill the queue in phase 1 with flows from which we can't deduce latency
(between A to  B). Such as a case in which we had HTTP traffic (A -> B) and
many many flows of DNS traffic (A -> C).

The new implementation uses a very large queue for 1st phase but then reduces
it down to a limitted set of unique flows: comparing ANodeTID, BNodeTID,
Application. As 2nd step is expenssive we only pass on a limitted list of 10
unique flows. So from the example above we would still see the HTTP traffic (A
-> B) becuase the DNS traffic (A -> C) would only use one spot in the list of
flows passed to the 2nd phase.

### CODE REVIEW

**Comment by safchain** [comment]
run cdd-overview

**Comment by safchain** [comment]
ok to test

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
run cdd-overview

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by hunchback** [comment]
@safchain - note I made changes such that the 2nd step executes only a single query ... so:
1st step - one query
2nd step - one query (optional)

File: statics/js/components/graph-layout.js [hunk 1]
```diff
@@ -94,21 +94,62 @@ var LinkLabelLatency = Vue.extend({
 
     flowQuery: function(nodeTID, trackingID, limit) {
       let query = `G.Flows()`;
-      query += `.Has("NodeTID", "${nodeTID}")`;
+      query += `.Has("NodeTID", ${nodeTID})`;
       if (typeof trackingID !== 'undefined') {
-        query += `.Has("TrackingID", "${trackingID}")`;
+        query += `.Has("TrackingID", ${trackingID})`;
```
Comment (safchain) [inline statics/js/components/graph-layout.js]: Would be better to use one Has in term of performances

File: statics/js/components/graph-layout.js [hunk 2]
```diff
@@ -94,21 +94,62 @@ var LinkLabelLatency = Vue.extend({
...
 
     flowQueryByNodeTID: function(nodeTID, limit) {
-      return this.flowQuery(nodeTID, undefined, limit);
+      return this.flowQuery(`"${nodeTID}"`, undefined, limit);
     },
 
-    flowQueryByNodeTIDandTrackingID: function(nodeTID, trackingID) {
-      return this.flowQuery(nodeTID, trackingID, 1);
+    flowQueryByNodeTIDandTrackingID: function(nodeTID, flows) {
+      anyTrackingID = 'Within(';
```
Comment (safchain) [inline statics/js/components/graph-layout.js]: "let" missing

**Comment by safchain** [comment]
rerun tests

File: statics/js/components/graph-layout.js (hunk 1 above)
Comment (hunchback) [reply statics/js/components/graph-layout.js]: ok, using one 'Has(...)'

File: statics/js/components/graph-layout.js (hunk 2 above)
Comment (hunchback) [reply statics/js/components/graph-layout.js]: added

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
run cdd-overview

**Pull request merged**